### ISSUE TITLE
The `object:` argument to `FieldExtension#resolve` does not have a consistent type

### ISSUE THREAD

**Issue by alexkalderimis** [rmosolgo/graphql-ruby#3385]
**Describe the bug**

Implementing `FieldExtension#resolve` is more difficult than it needs to be
because two different types of runtime values are passed to it, depending
on how the field is implemented.

If the field is defined with the inferred mechanisms (method, key-lookup on object)
or with a `Resolver`, then the object argument is a `Schema::Object` (outer-object).

If the field is defined with a resolver proc (or has been redefined
by field instrumentation to use a resolver proc) then the object argument
is a `Schema::Object#object` (inner-object).

Since procs are not-compatible with the interpreter, and not recommended, I assume
that the second of these two scenarios will go away at some point, but for now
`FieldExtension` implementations must handle both code-paths.

**Versions**

`graphql` version: At least up to 1.11.8, from looking at code it appears to
                 be the case on master (at 9b95a8bce5e92a2a531a1e015c9e15c549cd6cb7)
                 
`rails` (or other framework): Not framework dependent

**GraphQL schema**

Include relevant types and fields (in Ruby is best, in GraphQL IDL is ok).
Are you using [interpreter](https://graphql-ruby.org/queries/interpreter.html)? Any custom instrumentation, etc?

```ruby
class LoggingExtension < GraphQL::Schema::FieldExtension
  def resolve(object:, **)
    puts object.class
    yield(object)
  end
end
   
class Product < GraphQL::Schema::Object
  # assuming Product#object is a Hash
  field :id, ID, # logs Product
        null: false,
        hash_key: :id  
        extensions: [LoggingExtension]
  field :proc_id, ID, # logs Hash
        null: false,
        extensions: [LoggingExtension],
        resolve: ->(obj, args, ctx) { obj[:id] }
end
```

**Expected behavior**

The interface of `FieldExtension` should have consistent types.

**Actual behavior**

Implementing a `FieldExtension` has to account for code-paths that depend on internal
states outside the control of the field extension.

**Comment by alexkalderimis** [comment]
For reference the cause for this is the difference between https://github.com/rmosolgo/graphql-ruby/blob/9b95a8bce5e92a2a531a1e015c9e15c549cd6cb7/lib/graphql/schema/field.rb#L696 and https://github.com/rmosolgo/graphql-ruby/blob/9b95a8bce5e92a2a531a1e015c9e15c549cd6cb7/lib/graphql/schema/field.rb#L577

The first (the default resolution mechanism) calls the extension chain with `unextended_obj` which is an outer-object (i.e. `Schema::Object`).

The second calls the chain with `inner_object`, i.e. `Schema::Object#object` on the condition `if @resolve_proc`.

**Comment by rmosolgo** [comment]
Thanks for reporting this! I agree it's a bug. It _should_ always be "inner_object" -- I'm not sure how that slipped past 😖

**Comment by rmosolgo** [comment]
Unfortunately, I think I _have_ to leave this as-is. Now that I've released it this way, I don't want to break anyone who's already using it. I see that the issue pops up when using `resolve: -> ...`, but that isn't supported on the new runtime anyways. 

I added a test to keep an eye on this behavior in the future (d2be81c63). 

In the meantime, is there something you're trying to accomplish that this is blocking? I'd be happy to help think of a work-around if so.
